src/vos/vos_dtx.c

```diff
@@ -929,56 +948,55 @@ vos_dtx_register_record(struct umem_instance *umm, umem_off_t record,
 		struct vos_dtx_act_ent	*dae = dth->dth_ent;
 
 		/* Incarnation log entry implies a share */
-		*tx_id = dae->dae_df_off;
+		*tx_id = DAE_LID(dae);
 		if (type == DTX_RT_ILOG)
 			dth->dth_has_ilog = 1;
 	}
 
 	D_DEBUG(DB_IO, "Register DTX record for "DF_DTI
-		": entry %p, type %d, %s ilog entry, rc %d\n",
-		DP_DTI(&dth->dth_xid), dth->dth_ent, type,
-		dth->dth_has_ilog ? "has" : "has not", rc);
+		": lid=%d entry %p, type %d, %s ilog entry, rc %d\n",
+		DP_DTI(&dth->dth_xid),
+		DAE_LID((struct vos_dtx_act_ent *)dth->dth_ent), dth->dth_ent,
+		type, dth->dth_has_ilog ? "has" : "has not", rc);
 
 	return rc;
 }
 
 /* The caller has started PMDK transaction. */
 void
 vos_dtx_deregister_record(struct umem_instance *umm, daos_handle_t coh,
-			  umem_off_t entry, umem_off_t record)
+			  uint32_t entry, daos_epoch_t epoch, umem_off_t record)
 {
 	struct vos_container		*cont;
 	struct vos_dtx_act_ent		*dae;
 	struct vos_dtx_act_ent_df	*dae_df;
 	struct vos_dtx_record_df	*rec_df;
-	d_iov_t				 kiov;
-	d_iov_t				 riov;
+	bool				 found;
 	int				 count;
-	int				 rc;
 	int				 i;
 
-	if (!vos_dtx_is_normal_entry(umm, entry))
+	if (!vos_dtx_is_normal_entry(entry))
 		return;
 
-	dae_df = umem_off2ptr(umm, entry);
-	if (daos_is_zero_dti(&dae_df->dae_xid) ||
-	    dae_df->dae_flags & DTX_EF_INVALID)
-		return;
+	D_ASSERT(entry >= DTX_LID_RESERVED);
 
 	cont = vos_hdl2cont(coh);
 	if (cont == NULL)
-		goto handle_df;
+		return;
```

Comment:
Why not "goto handle_df"? For vos aggregation case, the "cont" is NULL, then how to handle the case of vos aggregation to remove the very old non-committed entry?

Reply:
why is cont NULL for aggregation?   I actually tried this but it seg faults because we can't actually get dae because it's in the cont array

Reply:
The main reason for NULL "coh" is related with dbtree_open_inplace_ex() called by gc_drain_btr(). In such case, the "coh" is set as DAOS_HDL_INVAL, that may cause subsequent ktr_rec_free() uses NULL "coh" to call vos_ilog_desc_cbs_init(), as to the vos_dtx_deregister_record() get NULL "coh".

Reply:
Here, if we return directly without cleanup related DTX entry (both in DRAM and on-disk), then when we call dtx_rec_release() to commit or abort such DTX, it will try to access related on-disk object/key that may have already been released or re-allocated for others, then cause data corruption.

Reply:
Maybe the gc should take a reference on the container then and save it to pass

Reply:
When gc handle related container, there is no in-DRAM container instance, because we are destroying related container. Then make gc to reference on the container seems not reasonable.

Reply:
Just realized that you may right, we can return directly if "cont" is NULL. Because it is for destroying container. Under such case, the in-DRAM DTX entry has been removed already. The on-disk entry will be destroyed soon. I will update the patch PR#2432 to do more cleanup.